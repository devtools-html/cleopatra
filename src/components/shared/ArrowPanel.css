/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

.arrowPanelAnchor {
  position: absolute;
  z-index: 10;
  left: 50%;
  top: 75%;
}

.arrowPanel {
  --internal-offset-from-right: var(--offset-from-right, 60px);
  --internal-offset-from-top: 15px;
  --internal-width: var(--width, initial);
  --internal-button-height: 30px;
  --internal-ok-button-background: var(--ok-button-background, #7990c8);
  --internal-ok-button-color: var(--ok-button-color, white);
  position: absolute;
  top: var(--internal-offset-from-top);
  right: calc(0px - var(--internal-offset-from-right));
  min-width: var(--internal-width);
  filter: drop-shadow(0 0 0.5px rgba(0,0,0,0.4)) drop-shadow(0 4px 5px rgba(0,0,0,0.4));
  filter: url(../../../res/img/svg/shadowfilter.svg#menushadow);
  color: black;
  text-align: left;
  line-height: 1.3;
  background: hsla(0, 0%, 97%, 0.95);
  border-radius: 5px;
  transform-origin: calc(100% - var(--internal-offset-from-right))
                    calc(0px - var(--internal-offset-from-top));
}

.arrowPanel:not(.open) {
  pointer-events: none;
  opacity: 0;
  /* visibility: hidden hides the panel from the accessibility tree when it's closed */
  visibility: hidden;
  /* The visibility is set to hidden after the panel closing transition. */
  transition: opacity 200ms ease-out, visibility 0s 200ms;
}

.arrowPanel.open {
  transition: none;
  animation: arrowPanelAppear 400ms;
  opacity: 1;
}

@keyframes arrowPanelAppear {
  from {
    opacity: 0;
    transform: scale(0);
    animation-timing-function: ease-out;
  }

  40% {
    opacity: 0.6;
    transform: scale(1);
    animation-timing-function: ease-in-out;
  }

  60% {
    opacity: 0.9;
    transform: scale(0.96);
    animation-timing-function: cubic-bezier(.3,0,0.3,1);
  }
}

.arrowPanelArrow {
  position: absolute;
  left: 0;
  right: 0;
  top: calc(0px - var(--internal-offset-from-top));
  height: var(--internal-offset-from-top);
  overflow: hidden;
}

.arrowPanelArrow::before {
  content: '';
  display: block;
  position: absolute;
  top: 0;
  left: calc(100% - var(--internal-offset-from-right));
  width: calc(1.42 * var(--internal-offset-from-top));
  height: calc(1.42 * var(--internal-offset-from-top));
  background: hsla(0, 0%, 97%, 0.95);
  transform: rotate(45deg);
  transform-origin: top left;
  z-index: -1;
}

.arrowPanelTitle,
.arrowPanelSubTitle {
  font-size: 1.2em;
  padding: 10px;
  margin: 0;
}

.arrowPanelSubTitle {
  border-bottom: 1px solid rgba(0,0,0,0.05);
  padding: 5px 0;
}

.arrowPanelContent {
  padding: 0 10px;
  overflow: auto;
  --internal-approx-distance-from-top: 60px;
  --internal-approx-distance-to-bottom: 100px;
  max-height: calc(100vh - var(--internal-approx-distance-from-top) - var(--internal-button-height) - var(--internal-approx-distance-to-top));
}

.arrowPanel.hasTitle > .arrowPanelContent {
  border-top: 1px solid rgba(0,0,0,0.05);
}

.arrowPanel.hasButtons > .arrowPanelContent {
  border-bottom: 1px solid rgba(0,0,0,0.05);
}

.arrowPanel ul {
  padding-left: 2em;
}

.arrowPanelButtons {
  height: 35px;
  line-height: 35px;
  display: flex;
  flex-flow: row nowrap;
  align-items: stretch;
}

.arrowPanelOkButton,
.arrowPanelCancelButton {
  border: 0;
  background: none;
  padding: 0;
  margin: 0;
  flex: 1;
  text-align: center;
  font-size: 1.2em;
}

.arrowPanelOkButton {
  border-bottom-right-radius: 5px;
  background: var(--internal-ok-button-background);
  color: var(--internal-ok-button-color);
}

.arrowPanelCancelButton {
  border-bottom-left-radius: 5px;
}

.arrowPanelOkButton:hover,
.arrowPanelCancelButton:hover {
  background-image: linear-gradient(rgba(0,0,0,0.1), rgba(0,0,0,0.1));
}

.arrowPanelOkButton:hover:active,
.arrowPanelCancelButton:hover:active {
  background-image: linear-gradient(rgba(0,0,0,0.2), rgba(0,0,0,0.2));
}

.arrowPanelOpenMetaInfo {
  width: 350px;
  font-size: 12px;
  line-height: 20px;
  padding-top: 5px;
  max-height: calc(100vh - 60px);
  overflow-y: auto;
}

.arrowPanelOpenMetaInfo,
.arrowPanelOpenMetaInfo .arrowPanelArrow::before {
  background-color: white;
}

.arrowPanelSection {
  margin-top: 7px;
  margin-bottom: 12px;
}

.metaInfoLabel {
  width: 120px;
  color: var(--grey-40);
  display: inline-block;
}

.metaInfoList {
  margin-top: -19px;
}

.metaInfoList .metaInfoListItem {
  margin-left: 110px;
  list-style: circle;
}
