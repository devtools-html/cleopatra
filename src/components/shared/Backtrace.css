.backtrace {
  --gradient-height: calc(var(--max-height) - 5em);

  /* Box */
  overflow: hidden;
  max-height: var(--max-height);
  padding: 0;
  margin: 5px;

  /* Other */
  font-size: 0.9em;
  line-height: 1.5;

  /* The gradient is defined towards the bottom so that we hide the end of the
   * backtrace only when the element is tall enough, which is when the backtrace
   * is long. */
  -webkit-mask-image: linear-gradient(
    to bottom,
    white,
    white var(--gradient-height),
    transparent
  );
  mask-image: linear-gradient(
    to bottom,
    white,
    white var(--gradient-height),
    transparent
  );
}

.backtraceStackFrame {
  /* Position */
  display: block;
  overflow: hidden;
  max-width: 35em;

  /* Box */
  padding: 0;
  margin: 0;

  /* Other */
  list-style: none;
  text-overflow: ellipsis;
  white-space: nowrap;
}

.backtraceStackFrameOrigin {
  margin-left: 10px;
  color: rgba(0, 0, 0, 0.4);
  font-style: normal;
}
