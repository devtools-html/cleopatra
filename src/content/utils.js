// @flow
type ValidWeakMapKey = Object | [];

/**
 * Memoize a function's results using a WeakMap and its first parameter.
 */
export function weakMemoize(fn: (ValidWeakMapKey, ...any) => any) {
  const memo = new WeakMap();
  return function (key: ValidWeakMapKey) {
    const cachedValue = memo.get(key);
    if (cachedValue === undefined) {
      const value = fn.apply(null, arguments);
      memo.set(key, value);
      return value;
    }
    return cachedValue;
  };
}
