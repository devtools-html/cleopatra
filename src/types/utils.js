/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow

/*
 * A not-void type
 * https://flow.org/try/#0C4TwDgpgBAcg9sAanAlgEygXigKClAHygDsBXAWwCMIAnPQqAZ2BpWIHN6jK44AbCAENiXEqT59RAQRo1BIADzCQAPlEB5SgCsIAY2ABuHDgBmpYvpRxiUMxYUAVAFywEydCoAUAN0F9SEC4OAJRBUADe9DQQwKQ0Nr7+EEYAvsZ2up4AjMFGGZ4ARMysHAW5puaZ4ba8LgDklII0dVAp5flkEu2VngDaWQA0UABMALrdFp7maBAmbBBoE5nlQA
 */
export type NotVoidOrNull = number | string | boolean | Array<any> | Object;

export type ExtractReturnType = <V>((...args: any[]) => V) => V;

/**
 * This type serves as documentation for how an array is meant to be used, but does
 * not support type checking. We often use an Array instead of a Map to translate
 * one type of index into another type of index. This is similar to how we use the
 * Map<K,V> type, but with the Array.
 */
// eslint-disable-next-line no-unused-vars
export type IndexedArray<_IndexType, Value> = Array<Value>;

/**
 * This is a utility type that extracts the return type of a function.
 */
export type $ReturnType<Fn> = $Call<ExtractReturnType, Fn>;
