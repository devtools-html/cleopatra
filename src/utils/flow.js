/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
// @flow

/**
 * This file contains utils that help Flow understand things better. Occasionally
 * statements can be logically equivalent, but Flow infers them in a specific way. Most
 * of the time tweaks can be done by editing the type system, but occasionally functions
 * are needed to get the desired result.
 */

/**
 * This function can be run as the default arm of a switch statement to ensure exhaustive
 * checking of a given type. It relies on an assumption that a given case will not be
 * the string 'Error: non exhaustive switch found.' This assumption generates more
 * readable errors. Flow will only generate an error if it's possible to get there
 * within the type system.
 */
export function unexpectedCase(notValid: empty): void {
  throw new Error(`Unexpected case ${notValid}`);
}

/**
 * Immutably update an object through Object.assign, but retain the original
 * type information of the object. Flow will occasionally throw errors when
 * inferring what is going on with Object.assign.
 */
export function immutableUpdate<T: Object>(object: T, ...rest: Object[]): T {
  return Object.assign({}, object, ...rest);
}
