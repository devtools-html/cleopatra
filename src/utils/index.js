/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
// @flow

/**
 * Firefox has issues switching quickly between fill style colors, as the CSS color
 * is fully parsed each time it is set. As a mitigation, provide a class that only
 * switches the color when it's really needed.
 */
export class FastFillStyle {
  _ctx: CanvasRenderingContext2D;
  _previousFillColor: string;

  constructor(ctx: CanvasRenderingContext2D) {
    this._ctx = ctx;
    this._previousFillColor = '';
  }

  set(fillStyle: string) {
    if (fillStyle !== this._previousFillColor) {
      this._ctx.fillStyle = fillStyle;
      this._previousFillColor = fillStyle;
    }
  }
}

/**
 * Perform a simple shallow object equality check.
 */
export function objectShallowEquals(a: Object, b: Object): boolean {
  let aLength = 0;
  let bLength = 0;
  for (const key in a) {
    if (a.hasOwnProperty(key)) {
      aLength++;
      if (a[key] !== b[key]) {
        return false;
      }
    }
  }
  for (const key in b) {
    if (b.hasOwnProperty(key)) {
      bLength++;
    }
  }
  return aLength === bLength;
}
