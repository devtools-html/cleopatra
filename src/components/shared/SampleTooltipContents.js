/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
// @flow

import * as React from 'react';

import { Backtrace } from './Backtrace';
import { getCategoryPairLabel } from 'firefox-profiler/profile-logic/profile-data';
import {
  formatNumber,
  formatMicroseconds,
  formatPercent,
} from 'firefox-profiler/utils/format-numbers';
import { ensureExists } from 'firefox-profiler/utils/flow';

import type {
  IndexIntoSamplesTable,
  CategoryList,
  Thread,
  SampleUnits,
  Milliseconds,
  ThreadCPUDeltaUnit,
} from 'firefox-profiler/types';

type Props = {|
  +sampleIndex: IndexIntoSamplesTable,
  +categories: CategoryList,
  +fullThread: Thread,
  +sampleUnits: SampleUnits | void,
  +maxThreadCPUDelta: number,
  +interval: Milliseconds,
|};

/**
 * This class displays the tooltip contents for a given sample. Typically the user
 * will want to know what the function is, and its category.
 */
export class SampleTooltipContents extends React.PureComponent<Props> {
  _getRealInterval(): number {
    const { fullThread, sampleIndex } = this.props;
    const { samples } = fullThread;

    // We don't know the first cpu usage so we can skip it.
    const index = sampleIndex === 0 ? 1 : sampleIndex;
    return samples.time[index] - samples.time[index - 1];
  }

  _getCPUPercentageString(
    cpuUsage: number,
    realInterval: Milliseconds
  ): string {
    const { sampleUnits, interval, maxThreadCPUDelta } = this.props;
    const unit = ensureExists(sampleUnits).threadCPUDelta;

    let cpuUsageRatio;
    let cpuSpikeText;
    switch (unit) {
      case 'variable CPU cycles': {
        // Here, we are finding the interval factor so we can find the CPU usage
        // per interval. This is similar to how we compute the max CPU.
        // cpuPerInterval makes us see the real percentage that we see in the
        // activity graph. Otherwise, the percentage here and in the activity
        // graph could be very different.
        const intervalFactor = realInterval / interval;
        const cpuPerInterval = cpuUsage / intervalFactor;
        cpuUsageRatio = cpuPerInterval / maxThreadCPUDelta;
        // We don't have a way to detect spikes for CPU cycles.
        cpuSpikeText = '';
        break;
      }
      case 'µs':
      case 'ns': {
        const maxCPU = realInterval * (unit === 'µs' ? 1000 : 1000000);
        cpuUsageRatio = cpuUsage / maxCPU;
        cpuSpikeText = cpuUsageRatio > 1 ? ', capped at 100%' : '';
        break;
      }
      default:
        throw new Error('Unexpected threadCPUDelta unit found');
    }

    return `${formatPercent(cpuUsageRatio)}${cpuSpikeText}`;
  }

  _getCPUUsageString(
    cpuUsage: number,
    cpuDeltaUnit: ThreadCPUDeltaUnit
  ): string {
    // It's either µs for timing platforms or 'variable CPU cycles' for Windows.
    const adjustedCPU = cpuDeltaUnit === 'ns' ? cpuUsage / 1000 : cpuUsage;
    const cpuUsageUnit = cpuDeltaUnit === 'ns' ? 'µs' : cpuDeltaUnit;
    return `${formatNumber(adjustedCPU, 2, 3)} ${cpuUsageUnit}`;
  }

  // Get thread CPU usage if it's present in the profile.
  _maybeRenderCpuUsage() {
    const { sampleIndex, fullThread, sampleUnits } = this.props;
    const { samples } = fullThread;
    let cpuUsageContent = null;

    if (sampleIndex === 0 || !samples.threadCPUDelta || !sampleUnits) {
      // We have no CPU usage information.
      return null;
    }

    const cpuUsage = samples.threadCPUDelta[sampleIndex];
    if (cpuUsage !== null) {
      const realIntervalMs = this._getRealInterval();
      const realIntervalUs = realIntervalMs * 1000;
      const percentageText = this._getCPUPercentageString(
        cpuUsage,
        realIntervalMs
      );
      const cpuUsageString = this._getCPUUsageString(
        cpuUsage,
        sampleUnits.threadCPUDelta
      );

      const cpuUsageAndPercentage = `${percentageText} (${cpuUsageString} over ${formatMicroseconds(
        realIntervalUs
      )})`;

      cpuUsageContent = (
        <div className="tooltipDetails">
          <div className="tooltipLabel">CPU Usage:</div>
          <div>{cpuUsageAndPercentage}</div>
        </div>
      );
    }

    return cpuUsageContent;
  }

  render() {
    const { sampleIndex, fullThread, categories } = this.props;
    const { samples, stackTable } = fullThread;
    const stackIndex = samples.stack[sampleIndex];
    if (stackIndex === null) {
      return 'No stack information';
    }
    const categoryIndex = stackTable.category[stackIndex];
    const subcategoryIndex = stackTable.subcategory[stackIndex];
    const categoryColor = categories[categoryIndex].color;

    return (
      <>
        <div className="tooltipDetails">
          <div className="tooltipLabel">Category:</div>
          <div>
            <span
              className={`colored-square category-color-${categoryColor}`}
            />
            {getCategoryPairLabel(categories, categoryIndex, subcategoryIndex)}
          </div>
        </div>
        {this._maybeRenderCpuUsage()}
        <div className="tooltipDetails">
          <div className="tooltipLabel">Stack:</div>
        </div>
        <Backtrace
          maxStacks={20}
          stackIndex={stackIndex}
          thread={fullThread}
          implementationFilter="combined"
          categories={categories}
        />
      </>
    );
  }
}
