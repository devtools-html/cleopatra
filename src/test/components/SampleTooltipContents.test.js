/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow

import React from 'react';
import { Provider } from 'react-redux';
import { SampleTooltipContents } from 'firefox-profiler/components/shared/SampleTooltipContents';
import { render, screen } from '@testing-library/react';
import {
  getCategories,
  getThreadSelectorsFromThreadsKey,
  getMaxThreadCPUDelta,
  getSampleUnits,
  getProfileInterval,
} from 'firefox-profiler/selectors';
import { ensureExists } from 'firefox-profiler/utils/flow';
import { storeWithProfile } from '../fixtures/stores';
import {
  getProfileFromTextSamples,
  addCpuUsageValues,
} from '../fixtures/profiles/processed-profile';

import type { Profile } from 'firefox-profiler/types';

describe('SampleTooltipContents', function() {
  function setup(profile: Profile, hoveredSampleIndex: number) {
    const store = storeWithProfile(profile);
    const state = store.getState();
    const threadSelectors = getThreadSelectorsFromThreadsKey(0);
    const fullThread = threadSelectors.getRangeFilteredThread(state);
    const categories = getCategories(state);
    const sampleUnits = getSampleUnits(state);
    const maxThreadCPUDelta = getMaxThreadCPUDelta(state);
    const interval = getProfileInterval(state);

    render(
      <Provider store={store}>
        <SampleTooltipContents
          sampleIndex={hoveredSampleIndex}
          fullThread={fullThread}
          categories={categories}
          sampleUnits={sampleUnits}
          maxThreadCPUDelta={maxThreadCPUDelta}
          interval={interval}
        />
      </Provider>
    );
  }

  it('renders the sample tooltip properly', () => {
    const { profile } = getProfileFromTextSamples(`
      _main
      XRE_main
      XREMain::XRE_main
      XREMain::XRE_mainRun
      nsAppStartup::Run
      nsAppShell::Run
      ChildViewMouseTracker::MouseMoved
      nsChildView::DispatchEvent
      nsView::HandleEvent
      nsViewManager::DispatchEvent
      mozilla::PresShell::HandleEvent
      mozilla::PresShell::HandlePositionedEvent
      mozilla::PresShell::HandleEventInternal
      mozilla::EventStateManager::PreHandleEvent
      mozilla::EventStateManager::GenerateMouseEnterExit
      mozilla::EventStateManager::NotifyMouseOut
      mozilla::EventStateManager::SetContentState
      mozilla::EventStateManager::UpdateAncestorState
      mozilla::dom::Element::RemoveStates
      nsDocument::ContentStateChanged
      mozilla::PresShell::ContentStateChanged
      mozilla::GeckoRestyleManager::ContentStateChanged
      mozilla::GeckoRestyleManager::PostRestyleEvent
      nsRefreshDriver::AddStyleFlushObserver[cat:Layout]
    `);
    // There is only one sample in the profile
    const hoveredSampleIndex = 0;

    setup(profile, hoveredSampleIndex);
    expect(document.body).toMatchSnapshot();
  });

  it('renders the sample with µs CPU usage information properly', () => {
    const { profile } = getProfileFromTextSamples(`
      A    A    A              A
      B    B    B              B
      Cjs  Cjs  H[cat:Layout]  H[cat:Layout]
      D    F    I[cat:Idle]
      Ejs  Ejs
    `);
    // Let's put some values for CPU usage.
    addCpuUsageValues(profile, [null, 400, 1000, 500], 'µs');

    // Let's check the second threadCPUDelta value
    const hoveredSampleIndex = 1;
    setup(profile, hoveredSampleIndex);

    const cpuUsage = ensureExists(
      screen.getByText(/CPU Usage/).nextElementSibling
    );
    expect(cpuUsage).toHaveTextContent('40% (400 µs over 1,000μs)');
    expect(document.body).toMatchSnapshot();
  });

  it('renders the sample with ns CPU usage information properly', () => {
    const { profile } = getProfileFromTextSamples(`
      A    A    A              A
      B    B    B              B
      Cjs  Cjs  H[cat:Layout]  H[cat:Layout]
      D    F    I[cat:Idle]
      Ejs  Ejs
    `);
    // Let's put some values for CPU usage.
    addCpuUsageValues(profile, [null, 600000, 1000000, 500000], 'ns');

    // Let's check the second threadCPUDelta value
    const hoveredSampleIndex = 1;
    setup(profile, hoveredSampleIndex);

    const cpuUsage = ensureExists(
      screen.getByText(/CPU Usage/).nextElementSibling
    );
    expect(cpuUsage).toHaveTextContent('60% (600 µs over 1,000μs)');
    expect(document.body).toMatchSnapshot();
  });

  it('renders the sample with "variable CPU cycles" CPU usage information properly', () => {
    const { profile } = getProfileFromTextSamples(`
      A    A    A              A
      B    B    B              B
      Cjs  Cjs  H[cat:Layout]  H[cat:Layout]
      D    F    I[cat:Idle]
      Ejs  Ejs
    `);
    // Let's put some values for CPU usage.
    addCpuUsageValues(profile, [null, 800, 900, 500], 'variable CPU cycles');

    // Let's check the second threadCPUDelta value
    const hoveredSampleIndex = 1;
    setup(profile, hoveredSampleIndex);

    const cpuUsage = ensureExists(
      screen.getByText(/CPU Usage/).nextElementSibling
    );
    expect(cpuUsage).toHaveTextContent(
      '89% (800 variable CPU cycles over 1,000μs)'
    );
    expect(document.body).toMatchSnapshot();
  });
});
