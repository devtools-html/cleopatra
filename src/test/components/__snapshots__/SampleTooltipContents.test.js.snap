// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SampleTooltipContents renders the sample tooltip properly 1`] = `
<body>
  <div>
    <div
      class="tooltipDetails"
    >
      <div
        class="tooltipLabel"
      >
        Category:
      </div>
      <div>
        <span
          class="colored-square category-color-purple"
        />
        Layout
      </div>
    </div>
    <div
      class="tooltipDetails"
    >
      <div
        class="tooltipLabel"
      >
        Stack:
      </div>
    </div>
    <ul
      class="backtrace"
    >
      <li
        class="backtraceStackFrame backtraceStackFrame_isFrameLabel"
      >
        <span
          class="colored-border category-color-purple"
          title="Layout"
        />
        nsRefreshDriver::AddStyleFlushObserver
        <em
          class="backtraceStackFrameOrigin"
        />
      </li>
      <li
        class="backtraceStackFrame backtraceStackFrame_isFrameLabel"
      >
        <span
          class="colored-border category-color-grey"
          title="Other"
        />
        mozilla::GeckoRestyleManager::PostRestyleEvent
        <em
          class="backtraceStackFrameOrigin"
        />
      </li>
      <li
        class="backtraceStackFrame backtraceStackFrame_isFrameLabel"
      >
        <span
          class="colored-border category-color-grey"
          title="Other"
        />
        mozilla::GeckoRestyleManager::ContentStateChanged
        <em
          class="backtraceStackFrameOrigin"
        />
      </li>
      <li
        class="backtraceStackFrame backtraceStackFrame_isFrameLabel"
      >
        <span
          class="colored-border category-color-grey"
          title="Other"
        />
        mozilla::PresShell::ContentStateChanged
        <em
          class="backtraceStackFrameOrigin"
        />
      </li>
      <li
        class="backtraceStackFrame backtraceStackFrame_isFrameLabel"
      >
        <span
          class="colored-border category-color-grey"
          title="Other"
        />
        nsDocument::ContentStateChanged
        <em
          class="backtraceStackFrameOrigin"
        />
      </li>
      <li
        class="backtraceStackFrame backtraceStackFrame_isFrameLabel"
      >
        <span
          class="colored-border category-color-grey"
          title="Other"
        />
        mozilla::dom::Element::RemoveStates
        <em
          class="backtraceStackFrameOrigin"
        />
      </li>
      <li
        class="backtraceStackFrame backtraceStackFrame_isFrameLabel"
      >
        <span
          class="colored-border category-color-grey"
          title="Other"
        />
        mozilla::EventStateManager::UpdateAncestorState
        <em
          class="backtraceStackFrameOrigin"
        />
      </li>
      <li
        class="backtraceStackFrame backtraceStackFrame_isFrameLabel"
      >
        <span
          class="colored-border category-color-grey"
          title="Other"
        />
        mozilla::EventStateManager::SetContentState
        <em
          class="backtraceStackFrameOrigin"
        />
      </li>
      <li
        class="backtraceStackFrame backtraceStackFrame_isFrameLabel"
      >
        <span
          class="colored-border category-color-grey"
          title="Other"
        />
        mozilla::EventStateManager::NotifyMouseOut
        <em
          class="backtraceStackFrameOrigin"
        />
      </li>
      <li
        class="backtraceStackFrame backtraceStackFrame_isFrameLabel"
      >
        <span
          class="colored-border category-color-grey"
          title="Other"
        />
        mozilla::EventStateManager::GenerateMouseEnterExit
        <em
          class="backtraceStackFrameOrigin"
        />
      </li>
      <li
        class="backtraceStackFrame backtraceStackFrame_isFrameLabel"
      >
        <span
          class="colored-border category-color-grey"
          title="Other"
        />
        mozilla::EventStateManager::PreHandleEvent
        <em
          class="backtraceStackFrameOrigin"
        />
      </li>
      <li
        class="backtraceStackFrame backtraceStackFrame_isFrameLabel"
      >
        <span
          class="colored-border category-color-grey"
          title="Other"
        />
        mozilla::PresShell::HandleEventInternal
        <em
          class="backtraceStackFrameOrigin"
        />
      </li>
      <li
        class="backtraceStackFrame backtraceStackFrame_isFrameLabel"
      >
        <span
          class="colored-border category-color-grey"
          title="Other"
        />
        mozilla::PresShell::HandlePositionedEvent
        <em
          class="backtraceStackFrameOrigin"
        />
      </li>
      <li
        class="backtraceStackFrame backtraceStackFrame_isFrameLabel"
      >
        <span
          class="colored-border category-color-grey"
          title="Other"
        />
        mozilla::PresShell::HandleEvent
        <em
          class="backtraceStackFrameOrigin"
        />
      </li>
      <li
        class="backtraceStackFrame backtraceStackFrame_isFrameLabel"
      >
        <span
          class="colored-border category-color-grey"
          title="Other"
        />
        nsViewManager::DispatchEvent
        <em
          class="backtraceStackFrameOrigin"
        />
      </li>
      <li
        class="backtraceStackFrame backtraceStackFrame_isFrameLabel"
      >
        <span
          class="colored-border category-color-grey"
          title="Other"
        />
        nsView::HandleEvent
        <em
          class="backtraceStackFrameOrigin"
        />
      </li>
      <li
        class="backtraceStackFrame backtraceStackFrame_isFrameLabel"
      >
        <span
          class="colored-border category-color-grey"
          title="Other"
        />
        nsChildView::DispatchEvent
        <em
          class="backtraceStackFrameOrigin"
        />
      </li>
      <li
        class="backtraceStackFrame backtraceStackFrame_isFrameLabel"
      >
        <span
          class="colored-border category-color-grey"
          title="Other"
        />
        ChildViewMouseTracker::MouseMoved
        <em
          class="backtraceStackFrameOrigin"
        />
      </li>
      <li
        class="backtraceStackFrame backtraceStackFrame_isFrameLabel"
      >
        <span
          class="colored-border category-color-grey"
          title="Other"
        />
        nsAppShell::Run
        <em
          class="backtraceStackFrameOrigin"
        />
      </li>
      <li
        class="backtraceStackFrame backtraceStackFrame_isFrameLabel"
      >
        <span
          class="colored-border category-color-grey"
          title="Other"
        />
        nsAppStartup::Run
        <em
          class="backtraceStackFrameOrigin"
        />
      </li>
      <span
        class="colored-border ellipsis"
      />
      …
    </ul>
  </div>
</body>
`;

exports[`SampleTooltipContents renders the sample with "variable CPU cycles" CPU usage information properly 1`] = `
<body>
  <div>
    <div
      class="tooltipDetails"
    >
      <div
        class="tooltipLabel"
      >
        Category:
      </div>
      <div>
        <span
          class="colored-square category-color-yellow"
        />
        JavaScript
      </div>
    </div>
    <div
      class="tooltipDetails"
    >
      <div
        class="tooltipLabel"
      >
        CPU Usage:
      </div>
      <div>
        89% (800 variable CPU cycles over 1,000μs)
      </div>
    </div>
    <div
      class="tooltipDetails"
    >
      <div
        class="tooltipLabel"
      >
        Stack:
      </div>
    </div>
    <ul
      class="backtrace"
    >
      <li
        class="backtraceStackFrame backtraceStackFrame_isFrameLabel"
      >
        <span
          class="colored-border category-color-yellow"
          title="JavaScript"
        />
        Ejs
        <em
          class="backtraceStackFrameOrigin"
        />
      </li>
      <li
        class="backtraceStackFrame backtraceStackFrame_isFrameLabel"
      >
        <span
          class="colored-border category-color-yellow"
          title="JavaScript"
        />
        F
        <em
          class="backtraceStackFrameOrigin"
        />
      </li>
      <li
        class="backtraceStackFrame backtraceStackFrame_isFrameLabel"
      >
        <span
          class="colored-border category-color-yellow"
          title="JavaScript"
        />
        Cjs
        <em
          class="backtraceStackFrameOrigin"
        />
      </li>
      <li
        class="backtraceStackFrame backtraceStackFrame_isFrameLabel"
      >
        <span
          class="colored-border category-color-grey"
          title="Other"
        />
        B
        <em
          class="backtraceStackFrameOrigin"
        />
      </li>
      <li
        class="backtraceStackFrame backtraceStackFrame_isFrameLabel"
      >
        <span
          class="colored-border category-color-grey"
          title="Other"
        />
        A
        <em
          class="backtraceStackFrameOrigin"
        />
      </li>
    </ul>
  </div>
</body>
`;

exports[`SampleTooltipContents renders the sample with ns CPU usage information properly 1`] = `
<body>
  <div>
    <div
      class="tooltipDetails"
    >
      <div
        class="tooltipLabel"
      >
        Category:
      </div>
      <div>
        <span
          class="colored-square category-color-yellow"
        />
        JavaScript
      </div>
    </div>
    <div
      class="tooltipDetails"
    >
      <div
        class="tooltipLabel"
      >
        CPU Usage:
      </div>
      <div>
        60% (600 µs over 1,000μs)
      </div>
    </div>
    <div
      class="tooltipDetails"
    >
      <div
        class="tooltipLabel"
      >
        Stack:
      </div>
    </div>
    <ul
      class="backtrace"
    >
      <li
        class="backtraceStackFrame backtraceStackFrame_isFrameLabel"
      >
        <span
          class="colored-border category-color-yellow"
          title="JavaScript"
        />
        Ejs
        <em
          class="backtraceStackFrameOrigin"
        />
      </li>
      <li
        class="backtraceStackFrame backtraceStackFrame_isFrameLabel"
      >
        <span
          class="colored-border category-color-yellow"
          title="JavaScript"
        />
        F
        <em
          class="backtraceStackFrameOrigin"
        />
      </li>
      <li
        class="backtraceStackFrame backtraceStackFrame_isFrameLabel"
      >
        <span
          class="colored-border category-color-yellow"
          title="JavaScript"
        />
        Cjs
        <em
          class="backtraceStackFrameOrigin"
        />
      </li>
      <li
        class="backtraceStackFrame backtraceStackFrame_isFrameLabel"
      >
        <span
          class="colored-border category-color-grey"
          title="Other"
        />
        B
        <em
          class="backtraceStackFrameOrigin"
        />
      </li>
      <li
        class="backtraceStackFrame backtraceStackFrame_isFrameLabel"
      >
        <span
          class="colored-border category-color-grey"
          title="Other"
        />
        A
        <em
          class="backtraceStackFrameOrigin"
        />
      </li>
    </ul>
  </div>
</body>
`;

exports[`SampleTooltipContents renders the sample with µs CPU usage information properly 1`] = `
<body>
  <div>
    <div
      class="tooltipDetails"
    >
      <div
        class="tooltipLabel"
      >
        Category:
      </div>
      <div>
        <span
          class="colored-square category-color-yellow"
        />
        JavaScript
      </div>
    </div>
    <div
      class="tooltipDetails"
    >
      <div
        class="tooltipLabel"
      >
        CPU Usage:
      </div>
      <div>
        40% (400 µs over 1,000μs)
      </div>
    </div>
    <div
      class="tooltipDetails"
    >
      <div
        class="tooltipLabel"
      >
        Stack:
      </div>
    </div>
    <ul
      class="backtrace"
    >
      <li
        class="backtraceStackFrame backtraceStackFrame_isFrameLabel"
      >
        <span
          class="colored-border category-color-yellow"
          title="JavaScript"
        />
        Ejs
        <em
          class="backtraceStackFrameOrigin"
        />
      </li>
      <li
        class="backtraceStackFrame backtraceStackFrame_isFrameLabel"
      >
        <span
          class="colored-border category-color-yellow"
          title="JavaScript"
        />
        F
        <em
          class="backtraceStackFrameOrigin"
        />
      </li>
      <li
        class="backtraceStackFrame backtraceStackFrame_isFrameLabel"
      >
        <span
          class="colored-border category-color-yellow"
          title="JavaScript"
        />
        Cjs
        <em
          class="backtraceStackFrameOrigin"
        />
      </li>
      <li
        class="backtraceStackFrame backtraceStackFrame_isFrameLabel"
      >
        <span
          class="colored-border category-color-grey"
          title="Other"
        />
        B
        <em
          class="backtraceStackFrameOrigin"
        />
      </li>
      <li
        class="backtraceStackFrame backtraceStackFrame_isFrameLabel"
      >
        <span
          class="colored-border category-color-grey"
          title="Other"
        />
        A
        <em
          class="backtraceStackFrameOrigin"
        />
      </li>
    </ul>
  </div>
</body>
`;
